api/watch/funcs_test.go

```diff
@@ -111,241 +99,310 @@ func TestKeyWatch(t *testing.T) {
 	c, s := makeClient(t)
 	defer s.Stop()
 
-	invoke := makeInvokeCh()
+	var (
+		wakeups  []*api.KVPair
+		notifyCh = make(chan struct{})
+	)
+
 	plan := mustParse(t, `{"type":"key", "key":"foo/bar/baz"}`)
 	plan.Handler = func(idx uint64, raw interface{}) {
-		if raw == nil {
-			return // ignore
-		}
-		v, ok := raw.(*api.KVPair)
-		if !ok || v == nil {
-			return // ignore
-		}
-		if string(v.Value) != "test" {
-			invoke <- errBadContent
-			return
+		var v *api.KVPair
+		if raw == nil { // nil is a valid return value
+			v = nil
+		} else {
+			var ok bool
+			if v, ok = raw.(*api.KVPair); !ok {
+				return // ignore
+			}
 		}
-		invoke <- nil
+
+		wakeups = append(wakeups, v)
+		notifyCh <- struct{}{}
 	}
 
 	var wg sync.WaitGroup
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		kv := c.KV()
+		if err := plan.Run(s.HTTPAddr); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+	}()
+	defer plan.Stop()
 
-		time.Sleep(20 * time.Millisecond)
+	// Wait for first wakeup.
+	<-notifyCh
+	{
+		kv := c.KV()
 		pair := &api.KVPair{
 			Key:   "foo/bar/baz",
 			Value: []byte("test"),
 		}
 		if _, err := kv.Put(pair, nil); err != nil {
 			t.Fatalf("err: %v", err)
 		}
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		if err := plan.Run(s.HTTPAddr); err != nil {
-			t.Fatalf("err: %v", err)
-		}
-	}()
-
-	if err := <-invoke; err != nil {
-		t.Fatalf("err: %v", err)
 	}
 
+	// Wait for second wakeup.
+	<-notifyCh
+
 	plan.Stop()
 	wg.Wait()
+
+	require.Len(t, wakeups, 2)
```

Comment:
Could this be racey since wakeups is being written to in another goroutine? Both in the sense of failing `-race` and in the sense of it's possible the length is now 3 presumably.

Reply:
We stop the plan and wait for it to stop running before reading `wakeups` so there shouldn't be a race.

```
$ go test ./api/watch/... -race
ok  	github.com/hashicorp/consul/api/watch	32.268s
```